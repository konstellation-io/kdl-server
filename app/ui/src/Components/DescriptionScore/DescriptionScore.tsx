import React, { memo, useEffect, useState } from 'react';
import { ScoreLevels, getScoreLevel, levels, scoreText } from './DescriptionScoreUtils';

import { SpinnerLinear } from 'kwc';
import cx from 'classnames';
import styles from './DescriptionScore.module.scss';

type Props = {
  score: number;
  loading?: boolean;
};

function DescriptionScore({ score, loading = false }: Props) {
  const [scoreLevel, setScoreLevel] = useState(ScoreLevels.LOW);
  const [scoreIndex, setScoreIndex] = useState(0);

  useEffect(() => {
    const selectedLevel = getScoreLevel(score);
    const selectedScoreIndex = levels.indexOf(selectedLevel);

    setScoreLevel(selectedLevel);
    setScoreIndex(selectedScoreIndex);
  }, [score]);

  function renderScoreLevels() {
    return levels.map((level, index) => (
      <div
        key={level}
        className={cx(styles.scoreLevel, {
          [styles[scoreLevel]]: !loading && index <= scoreIndex,
        })}
      />
    ));
  }

  function getScoreValue() {
    if (loading) return <SpinnerLinear size={40} />;

    return `${score}%`;
  }

  function getScoreDescription() {
    return loading ? 'Computing new description score...' : scoreText.get(scoreLevel);
  }

  return (
    <div className={styles.container}>
      <span className={cx(styles.scoreLabel, styles[scoreLevel])}>{getScoreValue()}</span>
      <div className={styles.scoreLevels}>{renderScoreLevels()}</div>
      <span className={styles.hintLabel}>{getScoreDescription()}</span>
    </div>
  );
}

export default memo(DescriptionScore);
