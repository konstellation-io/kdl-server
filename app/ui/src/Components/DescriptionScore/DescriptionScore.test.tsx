import DescriptionScore from './DescriptionScore';

import React from 'react';
import { mount } from 'enzyme';
import { getScoreLevel, ScoreLevels, scoreText } from './DescriptionScoreUtils';

// We have to use mount function of enzyme because of this https://enzymejs.github.io/enzyme/#react-hooks-support
// In this case it is not a problem use the mount method since DescriptionScore does not render any children
let wrapper;

beforeEach(() => {
  wrapper = mount(<DescriptionScore score={10} />);
});
describe('DescriptionScore component', () => {
  it('should render without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('behavior', () => {
    it('should show medium class if the score is between 25 and 50', () => {
      // Arrange.
      // Act.
      wrapper.setProps({
        score: 40,
      });
      wrapper.update();
      // Assert.
      expect(wrapper.exists(`.${ScoreLevels.MEDIUM}`)).toBe(true);
    });
    it('should show high class if the score is between 50 and 75', () => {
      // Arrange.
      // Act.
      wrapper.setProps({
        score: 55,
      });
      wrapper.update();
      // Assert.
      expect(wrapper.exists(`.${ScoreLevels.HIGH}`)).toBe(true);
    });
    it('should show very high class if the score is greater than 75', () => {
      // Arrange.
      // Act.
      wrapper.setProps({
        score: 80,
      });
      wrapper.update();
      // Assert.
      expect(wrapper.exists(`.${ScoreLevels.VERY_HIGH}`)).toBe(true);
    });
  });
});

describe('DescriptionScore utils', () => {
  describe('getScoreLevel', () => {
    it.each`
      score  | expectedLevel
      ${5}   | ${ScoreLevels.LOW}
      ${10}  | ${ScoreLevels.LOW}
      ${20}  | ${ScoreLevels.LOW}
      ${24}  | ${ScoreLevels.LOW}
      ${25}  | ${ScoreLevels.MEDIUM}
      ${30}  | ${ScoreLevels.MEDIUM}
      ${40}  | ${ScoreLevels.MEDIUM}
      ${45}  | ${ScoreLevels.MEDIUM}
      ${50}  | ${ScoreLevels.HIGH}
      ${55}  | ${ScoreLevels.HIGH}
      ${60}  | ${ScoreLevels.HIGH}
      ${70}  | ${ScoreLevels.HIGH}
      ${75}  | ${ScoreLevels.VERY_HIGH}
      ${80}  | ${ScoreLevels.VERY_HIGH}
      ${90}  | ${ScoreLevels.VERY_HIGH}
      ${100} | ${ScoreLevels.VERY_HIGH}
    `(
      'should return $expectedLevel when score is $score',
      ({ score, expectedLevel }) => {
        // Arrange.
        // Act.
        const result = getScoreLevel(score);
        // Assert.
        expect(result).toBe(expectedLevel);
      }
    );
  });

  describe('scoreText', () => {
    it('should match snapshot', () => {
      // Arrange.
      // Act.
      // Assert.
      expect(scoreText).toMatchSnapshot();
    });
  });
});
