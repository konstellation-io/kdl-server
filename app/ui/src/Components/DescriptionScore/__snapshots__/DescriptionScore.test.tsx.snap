// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DescriptionScore component should render without crashing 1`] = `
<Memo(DescriptionScore)
  score={10}
>
  <div
    className="container"
  >
    <span
      className="scoreLabel LOW"
    >
      10%
    </span>
    <div
      className="scoreLevels"
    >
      <div
        className="scoreLevel LOW"
        key="LOW"
      />
      <div
        className="scoreLevel"
        key="MEDIUM"
      />
      <div
        className="scoreLevel"
        key="HIGH"
      />
      <div
        className="scoreLevel"
        key="VERY_HIGH"
      />
    </div>
    <span
      className="hintLabel"
    >
      Description is incomplete for Knowledge Galaxy.
    </span>
  </div>
</Memo(DescriptionScore)>
`;

exports[`DescriptionScore utils scoreText should match snapshot 1`] = `
Map {
  "LOW" => "Description is incomplete for Knowledge Galaxy.",
  "MEDIUM" => "Please, write some more text in description.",
  "HIGH" => "Your description is good enough to be indexed.",
  "VERY_HIGH" => "Nice, your description is fully indexable.",
}
`;
